hydroflow_lang/src/graph/ops/halfjoin.rs

```diff
@@ -0,0 +1,143 @@
+///
+/// ```hydroflow
+/// // should print `(hello, (world, cleveland))`
+/// my_join = join();
+/// source_iter(vec![("hello", "world"), ("stay", "gold")]) -> [0]my_join;
+/// source_iter(vec![("hello", "cleveland")]) -> [1]my_join;
+/// my_join -> for_each(|(k, (v1, v2))| println!("({}, ({}, {}))", k, v1, v2));
+/// ```
+///
+/// `join` can also be provided with a single generic lifetime argument, either `'epoch` or
+/// `'static` to specify how join data persists. With `'epoch`, pairs will only be joined with
+/// corresponding pairs within the same epoch. With `'static`, pairs will be remembered across
+/// epochs and will be joined with pairs arriving in later epochs. When not explicitly specified
+/// persistence defaults to `static.
+///
+/// ```rustbook
+/// let (input_send, input_recv) = hydroflow::util::unbounded_channel::<(&str, &str)>();
+/// let mut flow = hydroflow::hydroflow_syntax! {
+///     my_join = join::<'epoch>();
+///     source_iter([("hello", "world")]) -> [0]my_join;
+///     source_stream(input_recv) -> [1]my_join;
+///     my_join -> for_each(|(k, (v1, v2))| println!("({}, ({}, {}))", k, v1, v2));
+/// };
+/// input_send.send(("hello", "oakland")).unwrap();
+/// flow.run_epoch();
+/// input_send.send(("hello", "san francisco")).unwrap();
+/// flow.run_epoch();
+/// ```
+/// Prints out "(hello, (world, oakland))" since `source_iter([("hello", "world")])` is only included
+/// in the first epoch, then forgotten.
+///
+/// ---
+///
+/// ```rustbook
+/// let (input_send, input_recv) = hydroflow::util::unbounded_channel::<(&str, &str)>();
+/// let mut flow = hydroflow::hydroflow_syntax! {
+///     my_join = join::<'static>();
+///     source_iter([("hello", "world")]) -> [0]my_join;
+///     source_stream(input_recv) -> [1]my_join;
+///     my_join -> for_each(|(k, (v1, v2))| println!("({}, ({}, {}))", k, v1, v2));
+/// };
+/// input_send.send(("hello", "oakland")).unwrap();
+/// flow.run_epoch();
+/// input_send.send(("hello", "san francisco")).unwrap();
+/// flow.run_epoch();
+/// ```
+/// Prints out "(hello, (world, oakland))" and "(hello, (world, san francisco))" since the inputs
+/// are peristed across epochs.
+#[hydroflow_internalmacro::operator_docgen]
+pub const JOIN: OperatorConstraints = OperatorConstraints {
+    name: "join",
+    hard_range_inn: &(2..=2),
+    soft_range_inn: &(2..=2),
+    hard_range_out: RANGE_1,
+    soft_range_out: RANGE_1,
+    ports_inn: Some(&(|| parse_quote! { 0, 1 })),
+    ports_out: None,
+    num_args: 0,
+    input_delaytype_fn: &|_| None,
+    write_fn: &(|wc @ &WriteContextArgs { root, op_span, .. },
+                 wi @ &WriteIteratorArgs { ident, inputs, .. },
+                 diagnostics| {
+        let persistence = parse_single_lifetime(wi, diagnostics).unwrap_or(Persistence::Static);
+
+        let joindata_ident = wc.make_ident("joindata");
+        let borrow_ident = wc.make_ident("joindata_borrow");
+        let (init, borrow) = match persistence {
```

Comment:
I confess I didn't descend into the sub-structures here to truly understand what you're doing.